src/plans/gradient_plan.jl

```diff
@@ -376,104 +457,114 @@ them to the current iterates tangent space.
+        vector_transport_to!(M, g, a.last_iterate, g, p, a.vector_transport_method)
     end
-    a.gradients[1] = deepcopy(d)
-    copyto!(M, a.last_iterate, get_iterate(s))
-    return step, 1 / length(a.gradients) .* sum(a.gradients)
+    pushfirst!(a.gradients, copy(M, p, d))
+    copyto!(M, a.last_iterate, p)
+    return 1.0, 1 / length(a.gradients) .* sum(a.gradients)
 end
 
-@doc raw"""
-    Nesterov <: DirectionUpdateRule
+"""
+    AverageGradient()
 
-## Fields
-* `γ`
-* `μ` the strong convexity coefficient
-* `v` (=``=v_k``, ``v_0=x_0``) an interim point to compute the next gradient evaluation point `y_k`
-* `shrinkage` (`= i -> 0.8`) a function to compute the shrinkage ``β_k`` per iterate.
+Add an average of gradients to a gradient processor. A set of previous directions (from the
+inner processor) and the last iterate are stored, average is taken after vector transporting
+them to the current iterates tangent space.
 
-Assume ``f`` is ``L``-Lipschitz and ``μ``-strongly convex. Given
+# Input
 
-* a step size ``h_k<\frac{1}{L}`` (from the [`GradientDescentState`](@ref)
-* a `shrinkage` parameter ``β_k``
-* and a current iterate ``x_k``
-* as well as the interim values ``γ_k`` and ``v_k`` from the previous iterate.
+* `M` (optional)
 
-This compute a Nesterov type update using the following steps, see [ZhangSra:2018](@cite)
+# Keyword arguments
 
-1. Compute the positive root ``α_k∈(0,1)`` of ``α^2 = h_k\bigl((1-α_k)γ_k+α_k μ\bigr)``.
-2. Set ``\bar γ_k+1 = (1-α_k)γ_k + α_kμ``
-3. ``y_k = \operatorname{retr}_{x_k}\Bigl(\frac{α_kγ_k}{γ_k + α_kμ}\operatorname{retr}^{-1}_{x_k}v_k \Bigr)``
-4. ``x_{k+1} = \operatorname{retr}_{y_k}(-h_k \operatorname{grad}f(y_k))``
-5. ``v_{k+1} = `\operatorname{retr}_{y_k}\Bigl(\frac{(1-α_k)γ_k}{\barγ_k}\operatorname{retr}_{y_k}^{-1}(v_k) - \frac{α_k}{\bar γ_{k+1}}\operatorname{grad}f(y_k) \Bigr)``
-6. ``γ_{k+1} = \frac{1}{1+β_k}\bar γ_{k+1}``
+* $(_kw_p_default)
+* `direction=`[`IdentityUpdateRule`](@ref) preprocess the actual gradient before adding momentum
+* `gradients=[zero_vector(M, p) for _ in 1:n]` how to initialise the internal storage
+* `n=10` number of gradient evaluations to take the mean over
+* $(_kw_X_default)`
+* `vector_transport_method=default_vector_transport_method(M, typeof(p)),
 
-Then the direction from ``x_k`` to ``x_k+1`` by ``d = \operatorname{retr}^{-1}_{x_k}x_{k+1}`` is returned.
+!!! info
+    This function generates a [`DirectionUpdateRuleFactory`](@ref) for this
+    for the [`AverageGradientRule`](@ref). If you do not provide the manifold
+    or any other of the keywords that require the manifold for a default,
+    you can provide a manifold later to use that manifold's default's instead.
+"""
+AverageGradient(args...; kwargs...) =
+    DirectionUpdateRuleFactory(AverageGradientRule, args...; kwargs...)
+
+@doc raw"""
+    NesterovRule <: DirectionUpdateRule
+
+Compute a Nesterov inspired direction update rule.
+See [`Nesterov`](@ref) for details
+
+# Fields
+
+* `γ`
```

Comment:
missing description after refactoring?

Reply:
Oh yes, I must have missed this.

In general since we now have 2 doc strings (one for the `XRule` one for `X`) I will split most docs into: The math part and how to use it will be in `X`, the technical details will be in the rule then.